[BUG] Using C++ API CMake

Hey guys,

First, loving the work, can't wait to get stuck in and have a play!

However I am having an issue building v0.0.4 when I use MLX as a dependency for my demo project?

I've tried to list everything here for you guys below.

Using CMake 3.28.3

Here is my CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.27)
project(a3c LANGUAGES CXX)

set(CMAKE_CXX_STANDARD 17)

include(FetchContent)
set(MLX_BUILD_METAL ON)

FetchContent_Declare(
        mlx
        GIT_REPOSITORY "https://github.com/ml-explore/mlx.git"
        GIT_TAG v0.4.0)
FetchContent_MakeAvailable(mlx)

add_executable(a3c src/main.cpp)
target_link_libraries(a3c PRIVATE mlx)
```

Configure output:
```
[main] Configuring project: a3c 
[proc] Executing command: /Users/cemlyn/conda/envs/a3c/bin/cmake --no-warn-unused-cli -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX:STRING=build -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_C_COMPILER:FILEPATH=/usr/bin/clang -DCMAKE_CXX_COMPILER:FILEPATH=/usr/bin/clang++ -S/Users/cemlyn/Code/GitHub/a3c -B/Users/cemlyn/Code/GitHub/a3c/build -G "Unix Makefiles"
[cmake] Not searching for unused variables given on the command line.
[cmake] -- Building MLX for arm64 processor on Darwin
[cmake] -- Building METAL sources
[cmake] -- Building with SDK for macOS version 14.2
[cmake] 
[cmake] -- Accelerate found /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/System/Library/Frameworks/Accelerate.framework
[cmake] -- Downloading json
[cmake] -- Using the multi-header code from /Users/cemlyn/Code/GitHub/a3c/build/_deps/json-src/include/
[cmake] -- Downloading gguflib
[cmake] CMake Deprecation Warning at build/_deps/doctest-src/CMakeLists.txt:1 (cmake_minimum_required):
[cmake]   Compatibility with CMake < 3.5 will be removed from a future version of
[cmake]   CMake.
[cmake] 
[cmake]   Update the VERSION argument <min> value or use a ...<max> suffix to tell
[cmake]   CMake that the project does not need compatibility with older versions.
[cmake] 
[cmake] 
[cmake] -- Configuring done (1.5s)
[cmake] -- Generating done (0.1s)
[cmake] -- Build files have been written to: /Users/cemlyn/Code/GitHub/a3c/build
```

Then building produces:

```
[main] Building folder: a3c a3c
[build] Starting build
[proc] Executing command: /Users/cemlyn/conda/envs/a3c/bin/cmake --build /Users/cemlyn/Code/GitHub/a3c/build --config Debug --target a3c -j 12 --
[build] make[3]: *** No rule to make target `/Users/cemlyn/Code/GitHub/a3c/mlx/types/half_types.h', needed by `_deps/mlx-build/mlx/backend/common/compiled_preamble.cpp'.  Stop.
[build] make[2]: *** [_deps/mlx-build/mlx/backend/common/CMakeFiles/cpu_compiled_preamble.dir/all] Error 2
[build] make[2]: *** Waiting for unfinished jobs....
[build] [  1%] Generating compiled_preamble.cpp
[build] [  2%] Building arange.air
[build] [  4%] Building steel_gemm.air
[build] [  5%] Building scatter.air
[build] [  6%] Building steel_gemm_splitk.air
[build] [  8%] Building C object _deps/mlx-build/mlx/io/CMakeFiles/gguflib.dir/__/__/__/gguflib-src/gguflib.c.o
[build] [  9%] Building softmax.air
[build] [  9%] Building steel_gemm_addmm.air
[build] [  9%] Building C object _deps/mlx-build/mlx/io/CMakeFiles/gguflib.dir/__/__/__/gguflib-src/fp16.c.o
[build] [ 10%] Building unary.air
[build] [ 12%] Building gather.air
[build] [ 13%] Building sort.air
[build] [ 13%] Built target compiled_preamble
[build] [ 14%] Building scan.air
[build] [ 14%] Building rope.air
[build] [ 16%] Building reduce.air
[build] [ 17%] Linking C static library libgguflib.a
[build] [ 18%] Building random.air
[build] [ 20%] Building quantized.air
[build] [ 20%] Built target gguflib
[build] [ 21%] Building gemv.air
[build] [ 21%] Building copy.air
[build] [ 22%] Building conv.air
[build] [ 24%] Building binary_two.air
[build] [ 25%] Building binary.air
[build] [ 26%] Building arg_reduce.air
[build] [ 26%] Building mlx.metallib
[build] [ 26%] Built target mlx-metallib
[build] make[1]: *** [CMakeFiles/a3c.dir/rule] Error 2
[build] make: *** [a3c] Error 2
[proc] The command: /Users/cemlyn/conda/envs/a3c/bin/cmake --build /Users/cemlyn/Code/GitHub/a3c/build --config Debug --target a3c -j 12 -- exited with code: 2
[driver] Build completed: 00:00:45.267
[build] Build finished with exit code 2
```

Note that for v0.3.0 it looks fine but obviously would love to use v0.4.0! ❤️

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -67,8 +67,6 @@ if (MLX_BUILD_METAL AND NOT METAL_LIB)
   set(MLX_BUILD_METAL OFF)
 elseif (MLX_BUILD_METAL)
   message(STATUS "Building METAL sources")
-  add_compile_definitions(_METAL_)
-
   # Throw an error if xcrun not found
   execute_process(COMMAND zsh "-c" "/usr/bin/xcrun -sdk macosx --show-sdk-version"
                   OUTPUT_VARIABLE MACOS_VERSION

--- a/mlx/backend/common/CMakeLists.txt
+++ b/mlx/backend/common/CMakeLists.txt
@@ -9,15 +9,15 @@ add_custom_command(
               ${CMAKE_CURRENT_SOURCE_DIR}/make_compiled_preamble.sh
               ${CMAKE_CURRENT_BINARY_DIR}/compiled_preamble.cpp
               ${CMAKE_CXX_COMPILER}
-              ${CMAKE_SOURCE_DIR}
+              ${PROJECT_SOURCE_DIR}
               ${CLANG}
 
     DEPENDS make_compiled_preamble.sh
             compiled_preamble.h
-            ${CMAKE_SOURCE_DIR}/mlx/types/half_types.h
-            ${CMAKE_SOURCE_DIR}/mlx/types/fp16.h
-            ${CMAKE_SOURCE_DIR}/mlx/types/bf16.h
-            ${CMAKE_SOURCE_DIR}/mlx/types/complex.h
+            ${PROJECT_SOURCE_DIR}/mlx/types/half_types.h
+            ${PROJECT_SOURCE_DIR}/mlx/types/fp16.h
+            ${PROJECT_SOURCE_DIR}/mlx/types/bf16.h
+            ${PROJECT_SOURCE_DIR}/mlx/types/complex.h
             ops.h
 )
 

--- a/mlx/backend/metal/CMakeLists.txt
+++ b/mlx/backend/metal/CMakeLists.txt
@@ -4,7 +4,7 @@ add_custom_command(
               ${CMAKE_CURRENT_SOURCE_DIR}/make_compiled_preamble.sh
               ${CMAKE_CURRENT_BINARY_DIR}/compiled_preamble.cpp
               ${CMAKE_C_COMPILER}
-              ${CMAKE_SOURCE_DIR}
+              ${PROJECT_SOURCE_DIR}
     DEPENDS make_compiled_preamble.sh
             kernels/compiled_preamble.h
             kernels/unary.h

--- a/mlx/backend/metal/metal.cpp
+++ b/mlx/backend/metal/metal.cpp
@@ -1,4 +1,4 @@
-// Copyright © 2023 Apple Inc.
+// Copyright © 2023-2024 Apple Inc.
 
 #include <cstdlib>
 #include <future>
@@ -9,6 +9,10 @@
 #include "mlx/scheduler.h"
 
 namespace mlx::core::metal {
+
+bool is_available() {
+  return true;
+}
 
 int max_ops_per_buffer() {
   auto get_val = []() {

--- a/mlx/backend/metal/metal.h
+++ b/mlx/backend/metal/metal.h
@@ -1,4 +1,4 @@
-// Copyright © 2023 Apple Inc.
+// Copyright © 2023-2024 Apple Inc.
 
 #pragma once
 
@@ -11,14 +11,7 @@
 
 namespace mlx::core::metal {
 
-constexpr bool is_available() {
-#ifdef _METAL_
-  return true;
-#else
-  return false;
-#endif
-}
-
+bool is_available();
 bool cache_enabled(void);
 void set_cache_enabled(bool enabled);
 

--- a/mlx/backend/no_metal/metal.cpp
+++ b/mlx/backend/no_metal/metal.cpp
@@ -1,10 +1,14 @@
-// Copyright © 2023 Apple Inc.
+// Copyright © 2023-2024 Apple Inc.
 
 #include <stdexcept>
 
 #include "mlx/backend/metal/metal.h"
 
 namespace mlx::core::metal {
+
+bool is_available() {
+  return false;
+}
 
 void new_stream(Stream) {}
 std::shared_ptr<void> new_scoped_memory_pool() {

